### DISCUSSION TITLE
What could be the reasons for un-necessary re-renders in React?

### DISCUSSION THREAD

**Discussion by kashif-hussain6** [vercel/next.js#69047] [App Router] [answered]
Unnecessary re-renders in React occur when components render more often than needed, which can slow down your application. This typically happens due to reasons like passing unchanged props, using inline functions, or not properly memoizing components. Minimizing these redundant re-renders is crucial for optimizing performance in React applications.

**Comment by Rashid-Aly** [comment, marked as answer]
Unnecessary re-renders in React can occur for several reasons:

State Changes: Updating the state without proper checks or conditions can trigger re-renders. Even if the state value remains the same, setting it again can cause a re-render.

Props Changes: Passing new references to props (like objects or arrays) even when their contents haven’t changed can lead to re-renders. This is because React treats different references as different values.

Parent Component Re-renders: If a parent component re-renders, all its child components will also re-render, even if the child props haven’t changed.

Inline Functions: Defining functions inline (e.g., inside the render method) creates a new function instance on each render, causing child components that use this function as a prop to re-render.

Context Changes: When context values change, all consuming components re-render, even if the change is not relevant to them.

Strict Mode: React’s Strict Mode intentionally causes double invocations of certain lifecycle methods, which might appear like unnecessary re-renders.

Using Non-Pure Components: Components that are not optimized with React.memo or shouldComponentUpdate can re-render unnecessarily.

Key Prop Misuse: Improper use of the key prop, especially in lists, can cause React to treat elements as different and re-render them.
